File: mobile/lib/core/api/auth_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config/app_config.dart';
import 'api_client.dart';

class AuthApi {
  /// Login vía backend (evita CORS Flutter Web → Keycloak).
  static Future<String> login(String username, String password) async {
    final user = username.trim();
    final pass = password;
    if (user.isEmpty || pass.isEmpty) {
      throw ApiException('Usuario y contraseña son obligatorios.');
    }

    http.Response response;
    try {
      response = await http.post(
        Uri.parse(AppConfig.loginUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'username': user, 'password': pass}),
      );
    } catch (_) {
      throw ApiException(
        'No se pudo conectar al servidor. Verifique backend (8080) y Keycloak (8180).',
      );
    }

    Map<String, dynamic>? data;
    try {
      data = jsonDecode(response.body) as Map<String, dynamic>?;
    } catch (_) {}

    if (response.statusCode != 200) {
      final msg = data?['message']?.toString() ?? 'Usuario o contraseña incorrectos.';
      throw ApiException(msg);
    }

    final token = data?['access_token'] as String?;
    if (token == null || token.isEmpty) {
      throw ApiException('No se recibió token de acceso.');
    }
    return token;
  }
}

File: frontend/src/shared/api/auth.api.ts
const KEYCLOAK_BASE =
  import.meta.env.VITE_KEYCLOAK_URL ??
  (import.meta.env.DEV
    ? '/realms/transporte-bluefields'
    : 'http://localhost:8180/realms/transporte-bluefields');

export async function loginKeycloak(username: string, password: string): Promise<string> {
  const user = username.trim();
  const pass = password;

  if (!user || !pass) {
    throw new Error('Usuario y contraseña son obligatorios.');
  }

  const body = new URLSearchParams({
    client_id: 'transporte-api',
    grant_type: 'password',
    username: user,
    password: pass,
  });

  let res: Response;
  try {
    res = await fetch(`${KEYCLOAK_BASE}/protocol/openid-connect/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body,
    });
  } catch {
    throw new Error(
      'No se pudo conectar con Keycloak. Verifique que Docker esté activo (puerto 8180) y reinicie el frontend.'
    );
  }

  if (!res.ok) {
    const err = await res.json().catch(() => null);
    const code = err?.error as string | undefined;
    if (code === 'invalid_grant') {
      throw new Error(
        'Usuario o contraseña incorrectos. Use el nombre de usuario (ej. admin.global), no el correo ni admin/admin de Keycloak. Contraseña demo: password'
      );
    }
    throw new Error(err?.error_description ?? err?.error ?? 'Credenciales inválidas');
  }

  return (await res.json()).access_token as string;
}
